### PR TITLE
Move `SECP256K1_INLINE` macro definition out from `include/secp256k1.h`

### PR DESCRIPTION
**Pull request by hebasto**
From [IRC](https://gnusha.org/secp256k1/2023-01-31.log):
> 06:29 \< hebasto\> What are reasons to define the `SECP256K1_INLINE` macro in user's `include/secp256k1.h` header, while it is used internally only?
> 06:32 \< hebasto\> I mean, any other (or a new dedicated) header in `src` looks more appropriate, no?
> 06:35 \< sipa\> I think it may just predate any "utility" internal headers.
> 06:42 \< sipa\> I think it makes sense to move it to util.h

Pros:
- it is a step in direction to better organized headers (in context of #924, #1039)

Cons:
- code duplication for `SECP256K1_GNUC_PREREQ` macro

### CODE REVIEW

**Comment by real-or-random** [comment]
Oh, the CI failures revealed a bug. The examples shouldn't use `SECP256K1_INLINE`.

**Comment by hebasto** [comment]
> Oh, the CI failures revealed a bug. The examples shouldn't use `SECP256K1_INLINE`.

Thanks! Fixed.

**Review by real-or-random — COMMENTED** [review]
Concept ACK

That's a breaking change technically, but in this case, even I tend not to treat it as one...  :see_no_evil: 

---

There's no strict reason why `util.h` shouldn't be able to include `secp256k1.h`, if it uses macro from the latter. If there's a reason to consider that a bad idea (but I don't think so),  and we prefer code duplication, we should probably add a comment to each copy, indicating that there's another copy that should be updated in lock-step.

File: src/util.h [hunk 1]
```diff
@@ -17,6 +17,27 @@
 #define DEBUG_CONFIG_MSG(x) "DEBUG_CONFIG: " x
 #define DEBUG_CONFIG_DEF(x) DEBUG_CONFIG_MSG(#x "=" STR(x))
 
+# if !defined(SECP256K1_GNUC_PREREQ)
+#  if defined(__GNUC__)&&defined(__GNUC_MINOR__)
```
[the remark this answers is not in the document]
Comment (hebasto) [reply src/util.h:21]: Cannot see any difference. What did I miss?

**Comment by hebasto** [comment]
> There's no strict reason why `util.h` shouldn't be able to include `secp256k1.h`...

Wouldn't that be exposing of implementation details in a public interface?

**Comment by real-or-random** [comment]
> > There's no strict reason why `util.h` shouldn't be able to include `secp256k1.h`...
> 
> Wouldn't that be exposing of implementation details in a public interface?

I meant adding `#include "../include/secp256k1.h"` to `util.h`, and not the other way around? Or how would that expose details?

File: src/util.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (real-or-random) [reply src/util.h:21]: Oh, you're right... I was looking at the missing whitespace around `&&`, which should be added. For some reason, I believed it's there in the public header, but I was probably looking at the wrong macro.

**Comment by hebasto** [comment]
> > > There's no strict reason why `util.h` shouldn't be able to include `secp256k1.h`...
> > 
> > 
> > Wouldn't that be exposing of implementation details in a public interface?
> 
> I meant adding `#include "../include/secp256k1.h"` to `util.h`, and not the other way around? Or how would that expose details?

You're right. I misunderstood you.

Well, it would work but it requires to add `include/` into include directories in build systems (`-I$(top_srcdir)/include` in Automake, `${PROJECT_SOURCE_DIR}/include` in CMake) like we do for examples now.

**Comment by hebasto** [comment]
> There's no strict reason why `util.h` shouldn't be able to include `secp256k1.h`, if it uses macro from the latter.

Thanks! Done.

**Pull request merged**